game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java

```diff
@@ -398,6 +380,18 @@ void addDependentUnits(final Map<Unit, Collection<Unit>> dependencies) {
     return new ArrayList<>(remaining);
   }
 
+  @Override
+  public List<Unit> getOffensiveAa() {
+    updateOffensiveAaUnits();
```

Comment:
This side effect is striking since getter methods tend to not mutate state. It could also be surprising if one caller were to use this class, and then we add another caller that then calls `getOffensiveAa` and *that* would change the state in a way the first caller might not be expecting.

Making the side-effect of setting the value of `offensiveAa` generally helps. IE:
```
offensiveAa = computeRemainingOffensiveAaUnits();
return offensiveAa;
```

That is still odd though, which makes me wonder if instead this problem would be solved by removing all side effects and instead calling the method something like `computeRemainingOffensiveAaUnits()`. In that case, any place where we call the getter and need those side effects, then we call the method instead. Perhaps this would lead to the `offensiveAa` variable to no longer be needed at all.

What are your thoughts?

Reply:
I agree it is surprising.  I had actually introduced a bug when I first started this because I assumed that `offensiveAa` was the value I wanted but it just so happened that `updateOffensiveAaUnits` was called earlier in the caller tree.  When I did some refactoring, my new code was called before `updateOffensiveAaUnits` and so I had the wrong set of aa units compared to the old code.

Unfortunately, `updateOffensiveAaUnits` actually updates 2 private values: `offensiveAa` and `offensiveAaTypes`.  And, as far as I can tell, they both need to stay in sync.

What I could do is to find the place early enough in the call chain where `offensiveAa` and `offensiveAaTypes` need to be initialized.  And then call `updateOffensiveAaUnits` there.  Then `getOffensiveAa` would just return the value.

Another method that has this side effect of updating offensiveAa is `canFireOffensiveAa`, by the way.  I hope to have remove that method when I convert the aa casualty removal step.

Reply:
Going over the code more, I think by calling `updateOffensiveAaUnits` every time `getOffensiveAa` is called might actually introduce a bug.  I think the `offensiveAa` variable should only be updated at the start of each battle round and then stay constant through the round.  So using a `computeRemainingOffensiveAaUnits()` might also introduce a bug.

I've explicitly called `updateOffensiveAaUnits` in all the places where it was implicitly being called originally so now it should be a little more obvious that it is being updated there.  At some future point, we can refactor it to be less brittle.